presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -53,11 +52,15 @@
     private final PlanNodeIdAllocator planNodeidAllocator = new PlanNodeIdAllocator();
     private final PlanVariableAllocator variableAllocator = new PlanVariableAllocator();
 
-    public static Optional<CanonicalPlanFragment> generateCanonicalPlan(PlanNode root, PartitioningScheme partitioningScheme)
```

Comment:
Could you reminds me a bit on why we need partitioning scheme at the very first place? For `outputLayout`, would that be the same as `canonicalPlan.get().getOutputVariables()`?

Reply:
We use partitioning scheme because if there is a mismatch between expected output layout and the output of the root node, we will add a layout enforce operation (e.g. https://github.com/prestodb/presto/blob/ab243b6ce24ac150bac4821c4a11342e89daa88a/presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java#L551). If we don't take this into consideration, then there might be channel mismatch which would lead to correctness issue or query failure. 

This `outputLayout` is not the same as `canonicalPlan.get().getOutputVariables()`, but rather the expected layout of the driver. At the beginning we are being conservative to ensure it is correct so we make sure everything in partitioning scheme is compared.

In fact, changing to `outputLayout` is not required for this PR, I could revert these changes and revisit this once we plan to support build side of join. What do you think?